fendermint/testing/materializer/tests/docker_tests/benches.rs

```diff
@@ -0,0 +1,494 @@
+        .await
+        .context("failed to get chain ID")?;
+
+    let cancel = Arc::new(Signal::new());
+
+    // Set up background blocks collector.
+    let blocks_collector = {
+        let cancel = cancel.clone();
+        let provider = provider.clone();
+        let assert = move |block: &Block<H256>| {
+            // Make sure block gas limit isn't the bottleneck.
+            let unused_gas_limit = block_gas_limit - block.gas_limit;
+            assert!(unused_gas_limit >= max_tx_gas_limit);
+        };
+        tokio::spawn(collect_blocks(cancel, provider, assert))
+    };
+
+    // Drive concurrency.
+    let cfg = Execution::new()
+        .add_step(1, 5)
+        .add_step(10, 5)
+        .add_step(100, 5)
+        .add_step(150, 5);
+    let testnet = Arc::new(testnet);
+    let testnet_clone = testnet.clone();
+    let nonce_manager = Arc::new(NonceManager::new());
+
+    let results = concurrency::execute(cfg.clone(), move |mut input| {
+        let testnet = testnet_clone.clone();
+        let nonce_manager = nonce_manager.clone();
+        let provider = provider.clone();
+
+        let test = async move {
+            let sender = testnet.account_mod_nth(input.test_id);
+            let recipient = testnet.account_mod_nth(input.test_id + 1);
+            println!("running (test_id={})", input.test_id);
+
+            let middleware = make_middleware(provider, sender, chain_id)
+                .await
+                .context("make_middleware")?;
+
+            let sender: H160 = sender.eth_addr().into();
+            let nonce = nonce_manager.get_and_increment(sender).await;
+
+            // Create the simplest transaction possible: send tokens between accounts.
+            let to: H160 = recipient.eth_addr().into();
+            let mut tx = Eip1559TransactionRequest::new()
+                .to(to)
+                .value(1)
+                .nonce(nonce);
+
+            let gas_estimation = middleware
+                .estimate_gas(&tx.clone().into(), None)
+                .await
+                .unwrap();
+            tx = tx.gas(gas_estimation);
+            assert!(gas_estimation <= max_tx_gas_limit);
+
+            input.bencher.start();
+
+            let pending: PendingTransaction<_> = middleware
+                .send_transaction(tx, None)
+                .await
+                .context("failed to send txn")?;
+            let tx_hash = pending.tx_hash();
+            println!("sent tx {:?} (test_id={})", tx_hash, input.test_id);
+
+            // We expect that the transaction is pending, however it should not return an error.
+            match middleware.get_transaction(tx_hash).await {
+                Ok(Some(_)) => {}
+                Ok(None) => bail!("pending transaction not found by eth hash"),
+                Err(e) => {
+                    bail!("failed to get pending transaction: {e}")
+                }
+            }
+            input.bencher.mempool();
+
+            let receipt = pending
```

Comment:
Maybe we should add a timeout here in case it isn't included?

Reply:
Yes, will be added.

Reply:
Added here: https://github.com/consensus-shipyard/ipc/pull/1243/commits/9c824183b9e31087bfa61b951b9af1454c4d7981